Carts and wishlists

MatteoPiovanelli-Laser opened the pull request:
Work on #117 
This PR contains the first portion of the work related to persistent carts and wishlists.

There is no wishlist functionality yet.

the ShoppingCart is treated as a content type with a ProductsListPart to store an authenticated user's cart in the item's infoset.

If a user already has a cart, we do not create a new one.

If a user is not logged in, their cart is stored as it was, using local storage and session storage.Once the user logs in, whatever they had in their carts gets merged into the persistent cart, and session and local storage are kept empty as long as the user is logged in.
To keep the local storage clear, I had to change the shoppingcart.js scripts a bit. Note that I did not change the minified version yet

Besides handling the peculiarity of login/logout scenarios, the cart works normally: items can be added/removed/updated, the cart is emptied on checkout and so on.

NOw I'll get to implementing the functionality of wishlists.

bleroy commented on Models/PersistentShoppingCart.cs:
```diff
@@ -0,0 +1,178 @@
+            var shoppingCartQuantities =
+                (from item in Items
+                 where productPartIds.Contains(item.ProductId) && item.Quantity > 0
+                 select new ShoppingCartQuantityProduct(item.Quantity, productParts.First(p => p.Id == item.ProductId), item.AttributeIdsToValues))
```
Please break that line down for readability.

MatteoPiovanelli-Laser replied:
ok

bleroy commented on Models/PersistentShoppingCart.cs:
```diff
@@ -0,0 +1,178 @@
+            return Math.Round(GetProducts().Sum(pq => Math.Round(pq.Price * pq.Quantity + pq.LinePriceAdjustment, 2)), 2);
+        }
+
+        public TaxAmount Taxes(decimal subTotal = 0) {
```
I'm thinking this implementation should be moved elsewhere, as it would be identical in any implementation of a cart.

bleroy replied:
Think about it this way: what is specific about this class, its single responsibility, should be persistence of the cart. More generic cart behavior should be moved.

MatteoPiovanelli-Laser replied:
created `public abstract class ShoppingCartBase`

bleroy commented on Models/PersistentShoppingCart.cs:
```diff
@@ -0,0 +1,178 @@
+                ).FirstOrDefault() ?? new TaxAmount { Amount = 0, Name = null };
+        }
+
+        public decimal Total(decimal subTotal = 0, TaxAmount taxes = null) {
```
This too. Maybe a base abstract class for default implementations would be useful here.

bleroy commented on Models/ProductsListPart.cs:
```diff
@@ -0,0 +1,74 @@
+
+        private List<ShoppingCartItem> _items {
+            get {
+                var fromInfoset = Retrieve<string>("SerializedItems");
```
It's wasteful to do this double serialization: serialize an array of objects into a string, that then gets serialized as part of the Infoset? You should be able to deal directly with the array without doing the serialization and deserialization yourself.

MatteoPiovanelli-Laser replied:
Using `Store` causes

```
Orchard.Exceptions.DefaultExceptionPolicy - NwazetTest - An unexpected exception was caught [https://localhost/Laser.Orchard/NwazetTest/addtocart?id=25]
System.NotSupportedException: Could not handle type List`1
```

MatteoPiovanelli-Laser replied:
Same thing if I use IEnumerable

MatteoPiovanelli-Laser replied:
I get that doing a first serialization to a string, and then storing that, is hacky (same stands for deserialization), but it's a simple workaround here.

bleroy replied:
That's why I said elsewhere that you don't necessarily have to go through the helpers. You can do what they do, but for your specific types.

bleroy commented on Models/ProductsListPart.cs:
```diff
@@ -0,0 +1,74 @@
+                return jsonInfoset["List"].Children()
+                    .Select(jt => {
+                        var item = new ShoppingCartItem(
+                            productId: int.Parse(jt["ProductId"].ToString()),
```
Converting to a string to immediately parse it back to an int? Why is this not already an int?

MatteoPiovanelli-Laser replied:
`jt["ProductId"]` here is a `Newtonsoft.Json.Linq.JToken`. It does not cast directly to int. Its `ToString()` gives em the string that in the json is the value for the parameter. THen I can parse that to int.

bleroy replied:
Try to cast it to `int?`. `Jtoken` has cast operators implemented.

bleroy commented on Models/ProductsListPart.cs:
```diff
@@ -0,0 +1,74 @@
+
+        public ShippingOption ShippingOption {
+            get {
+                var fromInfoset = Retrieve<string>("SerializedShippingOption");
```
Same thing here, the double serialization makes little sense.

bleroy replied:
Note that Store and Retrieve are not special, they are just helpers. You don't have to use them, necessarily.

MatteoPiovanelli-Laser replied:
fixed this. I did not know that store/retrieve where able to handle more complex objects.

MatteoPiovanelli-Laser replied:
Redacting: the ShippingOption has the same issue I have above, with `Store` not supporting it as a type.

bleroy replied:
Yes, you don't have to use the helpers, that are designed for simple types. Make your own.

bleroy commented on Module.txt:
```diff
@@ -95,4 +95,9 @@ Features:
+		Dependencies: Nwazet.AdvancedSKUManagement
+	Nwazet.PersistentCart:
+		Name: Nwazet Persistent Carts and WishLists
+		Description: Enables persistent (db) storage of user's shopping carts and wishlists
```
"Enable persistent database storage of [...]"

MatteoPiovanelli-Laser replied:
done

bleroy commented on Routes/CommerceRoutes.cs:
```diff
@@ -70,6 +70,20 @@ public class CommerceRoutes : IRouteProvider {
                         },
                         new MvcRouteHandler())
+                },
+                new RouteDescriptor {
```
Is a new route necessary, or would the default routes be sufficient?

MatteoPiovanelli-Laser replied:
THe default route does not seem to work. I am not really sure why, so I'll look into it, but I will keep the custom route as defined here for now.

MatteoPiovanelli-Laser replied:
Found the issue: it was in the javascript, where I would incorrectly use a token without checking whther it was not null.

MatteoPiovanelli-Laser commented on Drivers/ProductsListPartDriver.cs:
```diff
@@ -0,0 +1,14 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
will do.

MatteoPiovanelli-Laser commented:
- Removed unused using
 - Added `public abstract class ShoppingCartBase : IShoppingCart` with virtual implementations. `ShoppingCart` and `PersistentShoppingCart` inherit from that.
 - Edited description in module.txt. I removed the reference to wishlists as I am developing them in a separate feature.
 - Double serialization/deserialization: as I mentioned in my replies to your comments, the `Store` method seems to not accept the types here. I gave a quick look in the sources, and I think the issue lies with the `XmlHelper.ToString<T>` method. FOr now, I am keeping the double serialization using Newtonsoft.Json.

An update to the PR is coming soon.

MatteoPiovanelli-Laser commented:
Hi,

I am back on this after the past couple of weeks working on the concurrency issues of Orchard.

I wrote the serialization.
I preferred the version that was serializing twice, honestly, because the code was more compact, and it was less likely to require updates in case the classes get changed.

MatteoPiovanelli-Laser commented:
There are issues in this PR with the scripts. To reproduce them:

Have a cart with 2+ different items.
Open a page that shows the cart widget.
Hit "remove" for one item (this works).
Hit "remove for the second item (this does notcall the controller).

Likewise, "Update all" works only once.

I probably messed something up when I updated the scripts to handle local storage differently.

bleroy commented on Models/ProductsListPart.cs:
```diff
@@ -1,35 +1,90 @@
+        private IEnumerable<ShoppingCartItem> RetrieveItems() {
+            var partElement = GetInfosetElement();
+            return partElement.Elements()
+                .Where(el => el.Name.LocalName.StartsWith("ShoppingCartItem_")) //XElements corresponding to items
```
Why isn't `ShoppingCartItem` an array, and `AttributeIdsToValues` an object?

bleroy replied:
I think those should be children of a `ShoppingCartItems` element, not multiple elements with generated tag names.

bleroy commented on Models/ProductsListPart.cs:
```diff
@@ -1,35 +1,90 @@
+                        productId: int.Parse(el.Attribute("ProductId").Value),
+                        quantity: int.Parse(el.Attribute("Quantity").Value),
+                        attributeIdsToValues: el.Elements()
+                            .Where(subel => subel.Name.LocalName.StartsWith("AttributeIdsToValues_")) //XElements corresponding to dictionary entries
```
Why do these need to have different tag names? Looks like child elements would work better and be easier.

bleroy commented on Models/ProductsListPart.cs:
```diff
@@ -49,11 +104,74 @@ public class ProductsListPart : ContentPart {
+            }
+        }
+        private string ToStringAndLength(string value) {
+            return string.Format(@"{{{0}}}{{{1}}}", value.Length, value);
```
Just concatenate: `String.Format` adds no value here.

bleroy commented on Models/ProductsListPart.cs:
```diff
@@ -49,11 +104,74 @@ public class ProductsListPart : ContentPart {
+                FormValue = shipElement.Attribute("FormValue").Value
+            } : null;
+        }
+        private IEnumerable<string> FromStringAndLength(string serialized) {
```
I don't understand what you're doing here or why it's necessary.

bleroy replied:
It looks like you came up with your own way to serialize an array of strings. There's a perfectly valid way of doing that in XML:

```xml
<sometag>
  <someOtherTag>value 1</someOtherTag>
  <someOtherTag>value 2</someOtherTag>
</sometag>
```

MatteoPiovanelli-Laser commented:
I think I made the changes per your review of the serialization part

I fixed the bug with shoppingcart.js

bleroy commented on Models/ProductsListPart.cs:
```diff
@@ -28,23 +28,27 @@ public class ProductsListPart : ContentPart {
-                                };
-                                return new KeyValuePair<int, ProductAttributeValueExtended>(id, val);
-                            }
+                        productId: int.Parse(itemElement.Attribute("ProductId").Value),
```
I should have spotted this one earlier: you can actually cast `XElement` to `int`: https://msdn.microsoft.com/en-us/library/bb348319(v=vs.110).aspx

bleroy commented:
Thanks, I think this looks pretty good. I have one final comment on parsing ints from XML.

MatteoPiovanelli-Laser commented:
I updated the branch.

A few considerations:

- `ProductsListPart` has a `bool IsCart` property. This is of course `true` for carts. However that property is currently not used anywhere else. For wishlists, I am doing things slightly different, as I outlined in #117.
- Still regarding wishlists, I am working on a separate feature that does not depend on this one.
- Later I will need to add the events and activities we discussed that allow the creation of workflows to track the cart history. I will get on preparing a separate PR once this feature is "finalized".

bleroy commented:
Thanks!
